operator/pkg/livestate/livestate.go

```diff
@@ -0,0 +1,60 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package livestate
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/GoogleCloudPlatform/k8s-config-connector/operator/pkg/apis/core/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/operator/pkg/k8s"
+)
+
+// FetchLiveKCCState tries to fetch the ConfigConnector (CC) resource and the ConfigConnectorContext (CCC)
+// for the resource's namespace if running in Namespaced mode. It ignores not found errors for CC fetching
+// but errors out if KCC is running in Namespaced mode and no CCC is found for the namespace of the resource.
+func FetchLiveKCCState(ctx context.Context, c client.Client, resourceNN types.NamespacedName) (v1beta1.ConfigConnector, v1beta1.ConfigConnectorContext, error) {
+	var cc v1beta1.ConfigConnector
+	if err := c.Get(ctx, types.NamespacedName{
+		Name: k8s.ConfigConnectorAllowedName,
+	}, &cc); err != nil {
+		if errors.IsNotFound(err) {
+			// if no CC exists, then by definition, KCC cannot be running in namespaced mode;
+			return v1beta1.ConfigConnector{}, v1beta1.ConfigConnectorContext{}, nil
+		}
+		return v1beta1.ConfigConnector{}, v1beta1.ConfigConnectorContext{}, err
+	}
+
+	if cc.Spec.Mode == k8s.NamespacedMode {
+		var ccc v1beta1.ConfigConnectorContext
+		if err := c.Get(ctx, types.NamespacedName{
+			Name:      k8s.ConfigConnectorContextAllowedName,
+			Namespace: resourceNN.Namespace,
+		}, &ccc); err != nil {
+
```

Comment:
I really think we shall start using klogs in KCC. (or maybe we did already ? ) 😃

Reply:
We have klog in top level entry points like the controllers (e.g. https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/pkg/controller/tf/controller.go#L62). Though we don't have them in files/code for detailed logic.

The errors in the detailed logic will be surfaced back as the resource status, and too much details in the log can be distractive in debugging and can slow down the test log rendering. But I'm not strongly opinionated.

@yuwenma feel free to create a GitHub issue about it and prioritize it accordingly!

Reply:
As for "too much log details", that's what [`klog.V`](https://pkg.go.dev/k8s.io/klog/v2) can help with. 
In general, we want to log special handlings and abnormal behaviors. They are not only for debugging, but help us evaluate the performance and any potential issues.
 
I think we don't need to add klogs to all the existing code, since that code has run for a long time and should work well. But for new features, especially those that change the KCC behaviors or could cause breakages,  we will want to keep a closer eye on. I think stateintoSpec override falls into that category.

Reply:
Thank you for the suggestions, @yuwenma ! I created #1588 to follow up. I think adding klog can be a good enhancement to the code and I'll address it separately. Meanwhile, I want to prioritize merging this PR as it's been hanging for several weeks.

Reply:
Fair enough! Let's merge it and do incremental improvements!